cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+      }
+      return list;
+
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the metadata for the element identified by the given name.
+   * @param name
+   * @return
+   * @throws IOException
+   */
+  SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    if (cache.containsKey(metaKey)) {
+      return cache.get(metaKey);
+    }
+    try {
+      if (!keyStore.containsAlias(metaKey)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      SecureStoreMetadata meta = ((KeyMetadata) keyStore.getKey(metaKey, password)).metadata;
+      cache.put(metaKey, meta);
+      return meta;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name);
+    }
+  }
+
+  /**
+   * Returns the securely stored data as a UTF8 encoded byte array.
+   * @param name
+   * @return
+   * @throws IOException
+   */
+  byte[] getData(String name) throws IOException {
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      SecretKeySpec key = (SecretKeySpec) keyStore.getKey(name, password);
+      return key.getEncoded();
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the key " + name);
+    }
+  }
+
+  /**
+   * @return the provider configuration
+   */
+  Configuration getcConf() {
+    return cConf;
+  }
+
+  private static Path constructOldPath(Path path) {
+    return Paths.get(path.toString(), "_OLD");
+  }
+
+  private static Path constructNewPath(Path path) {
+    return Paths.get(path.toString(), "_NEW");
+  }
+
+  private static void loadFromPath(KeyStore keyStore, Path path, char[] password)
+    throws IOException {
+    try (InputStream in = new DataInputStream(Files.newInputStream(path))) {
+      keyStore.load(in, password);
+    } catch (NoSuchAlgorithmException | CertificateException e) {
+      throw new IOException("Unable to load the Secure Store. ", e);
+    }
+  }
+
+  private static boolean isBadOrWrongPassword(IOException ioe) {
+    /* According to Java keystore documentation if the load failed due to bad password
+       then the cause of the exception would be set to "UnrecoverableKeyException".
+       Unfortunately that is not the observed behavior. */
+    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+      return true;
+    }
+    // Workaround
+    return (ioe.getCause() == null)
+      && (ioe.getMessage() != null)
+      && ((ioe.getMessage().contains("Keystore was tampered")) || (ioe
+      .getMessage().contains("password was incorrect")));
+  }
+
+  private static void renameOrFail(Path src, Path dest) throws IOException {
```

Comment:
in fact, do we actually need this method? Seems like caller are always providing absolute path?